src/libasr/codegen/llvm_utils.cpp

```diff
@@ -1564,6 +1564,12 @@ namespace LCompilers {
                                         llvm_utils->create_ptr_gep(key_mask, key_hash));
         llvm::Value* is_prob_not_neeeded = builder->CreateICmpEQ(key_mask_value,
             llvm::ConstantInt::get(llvm::Type::getInt8Ty(context), llvm::APInt(8, 1)));
+        std::pair<std::string, std::string> llvm_key = std::make_pair(
+            ASRUtils::get_type_code(key_asr_type),
+            ASRUtils::get_type_code(value_asr_type)
+        );
+        llvm::Type* value_type = std::get<2>(typecode2dicttype[llvm_key]).second;
+        llvm::Value* result = builder->CreateAlloca(value_type, nullptr);
```

Comment:
I think `result` should only be allocated when `def_value` is not `nullptr`. I think you should not mix `read_item` with the logic of `dict.get`. Make a new method `get_item` and call it to implement `dict.get`. Direct access via `[]` should be done by `read_item` (which should be left untouched in this PR).

Reply:
Presently, `[]` and `dict.get` are implemented with the same `DictItem` node in the ASR and we can re-use the same implementation just to extend it with default. I am not sure if it is a good idea to re-implement the same logic for `get_item` and `read_item` which effectively are doing the same job.

> I think `result` should only be allocated when `def_value` is not `nullptr`. 

It's also used on line 1597 which also makes it more generalized in the case when key_hash is matching but not the keys. This was the bug with the previous implementation where it raised an error for the case where key_hash matches but not the key while it didn't raise any error when the key is totally absent. This is also fixed using the current PR. So, I don't think separating two methods for implementing same logic is a good idea.

Reply:
Well in this PR itself we have seen (as pointed in https://github.com/lcompilers/lpython/pull/1700#discussion_r1169024218 and https://github.com/lcompilers/lpython/pull/1700#issuecomment-1512464866) how `read_item` is also affected when you implement `dict.get` inside it. So de-coupling `if-else-if` checks 



> I am not sure if it is a good idea to re-implement the same logic for `get_item` and `read_item` which effectively are doing the same job.

If code repetition is the issue then you can solve it by using a macro, or a utility method and call it from `read_item` and `get_item`.



> Presently, `[]` and `dict.get` are implemented with the same `DictItem` node in the ASR and we can re-use the same implementation just to extend it with default

Won't be the case very soon. We will use `IntrinsicFunction` to implement, `dict.item` and `dict.get` separately.

> It's also used on line 1597 which also makes it more generalized in the case when key_hash is matching but not the keys.

In https://github.com/lcompilers/lpython/blob/c935c96b50abeb0553213904a9f96247f6f78421/src/libasr/codegen/llvm_utils.cpp, AFAICT, its only used when `def_value != nullptr`,

https://github.com/lcompilers/lpython/blob/c935c96b50abeb0553213904a9f96247f6f78421/src/libasr/codegen/llvm_utils.cpp#L1596-L1598

Again, my POV is still the same `read_item` and `get_item` should be different methods. However they can call common utility methods or macros to execute the common logic. `if (some_arg_is_present)` and then `else` always indicate that de-coupling is needed.



> This was the bug with the previous implementation where it raised an error for the case where key_hash matches but not the key while it didn't raise any error when the key is totally absent.

Please add a test for it and fix the `read_item` and `get_item` methods accordingly.